### DISCUSSION TITLE
Any options for when a belief addition plan fails?

### DISCUSSION THREAD

**Discussion by mindcrime** [jason-lang/jason#109] [Q&A] [answered]
I was searching the 'net and found [this StackOverflow question](https://stackoverflow.com/questions/68088891/how-to-execute-a-second-plan-when-the-first-fail-in-qwery-in-jason-agnetspeak) which sort-of answers my question, which amounts to "how can you handle failure in a plan triggered by the addition of a belief?" I see the answer there provided by Jomi says pretty simply 

```
failure plans can be used only for goals
```

So my follow-up question is just "is there *any* option for handling failure in such plans, or is this just  not possible?" And if failure is a possibility is the canonical pattern to have the +belief plan just add an achievement goal, and then use the normal failure handling mechanism for handling goal failures? 

And I guess a follow-on question might be, is it generally the case that most plans should probably be triggered by achievement goals? Is that the more idiomatic way to proceed? 

Thanks again!

**Comment by jomifred** [comment, marked as answer]
Hi, that is a good point to discuss. 

AFAIK, there is no "oficial" proposal for that. The problem, however, is known and has motivated some publications (http://dx.doi.org/10.1007/978-3-030-25693-7_3).

Personally, I prefer to program goal oriented. So, belief events quickly trigger goals. Among the advantages: 
  - failure handling
  - all actions are explained by (can be assigned to) a goal

Jomi

**Reply by jomifred** [reply]
Briefly, in the cited paper, we propose some extensions for AgentSpeak that allow us to write the following code

```
+!start <: false. {
    +oi <- act.
    -!start <- .println(ok).
}
```

- the goal `start` will never be accomplished (what follows `<:` is the drop condition for the goal).
- the plan for `+oi` (a sub-plan of start) is 'available' only when the agent is pursuing goal `start`. So, by having the goal `start`, the agent will react to some events.
- the action `act` is thus motivated by goal `start` (the idea the all actions are due to some goal)
- in case of failure of action `act`, the plan for `-!start` is used

Sub-plans is available since Jason 3.2. However, the failure handling feature will be available in Jason 3.3.

**Reply by mindcrime** [reply]
Thanks Jomi. Sounds like at least for now the thing to do is focus on using goals as you say. I will review the linked paper as well though, and look forward to any changes in Jason/AgentSpeak that come down the line.

**Comment by rbordini** [comment]
Indeed an interesting discussion. I agree with Jomi that agents should
normally react to external events by adopting new goals.

**Reply by mindcrime** [reply]
Makes sense. To be honest, most of my use of "belief addition plans" so far has just been for "hello world" type scenarios. But it makes sense that goals should drive most behavior in "real" agents.  :-)
